### DISCUSSION TITLE
Improving syntax highlighting

### DISCUSSION THREAD

**Discussion by joshtemple** [tconbeer/harlequin#444] [Ideas (Feature Requests)]
Hey @tconbeer! Do you have any plans for Harlequin to allow adapters to syntax highlight the SQL using their own dialect?

I've noticed, at least for BigQuery SQL, the syntax isn't correctly highlighted due to dialect-specific differences.

![image](https://github.com/tconbeer/harlequin/assets/8672171/3201fe75-9074-4c48-b1d6-994c8ca78d99)

**Comment by tconbeer** [comment]
yes, the syntax highlighting kinda sucks right now. There is an open issue for one solution for this (#340 ), which would be adapter-specific, but I also want to improve the default experience.

We're currently using [tree-sitter](https://github.com/grantjenks/py-tree-sitter-languages) for query parsing and syntax highlighting. It seems like [`tree-sitter-sql`](https://github.com/m-novikov/tree-sitter-sql) has a pretty complete grammar, but a very basic set of [queries for highlights](https://github.com/m-novikov/tree-sitter-sql/blob/main/queries/highlights.scm). It's theoretically possible to define and build your own tree sitter language, but it's much easier to define your own highlight query (Textual is pluggable for either, but a language requires building JS to C and then compiling C to binaries on tons of platforms). 

There is also [tree-sitter-sqlite](https://github.com/dhcmrlchtdj/tree-sitter-sqlite/blob/main/queries/highlights.scm), but those highlight queries seem equally basic. There is also [tree-sitter-sql-bigquery](https://github.com/takegue/tree-sitter-sql-bigquery), but that language isn't distributed with `py-tree-sitter-languages`, and it might be hard to distribute it with Harlequin (I really don't know). 

All that said, my current plan of attack is to try to write a better highlight query that would work for many databases; we could simply distribute that query source file with Harlequin, so we don't have to worry about builds. I haven't started that work.

**Comment by joshtemple** [comment]
@tconbeer if I wanted to update the BigQuery adapter to use the `tree-sitter-sql-bigquery` package you mentioned above for syntax highlighting, would Harlequin support the adapter overriding it?
